### ISSUE TITLE
there is no code, if you have, we had better learn your inspired code

### ISSUE THREAD

**Issue by githubgtl** [lsvih/SOAE#1]
there is no code, if you have, we had better learn your inspired code

**Comment by lsvih** [comment]
Hi there, you can follow the setting from [TPLinker](https://github.com/131250208/TPlinker-joint-extraction) and transfer our dataset to the corresponding data struct by referring to code below:

```python
import json
import logging
import os
import re
import unicodedata
from pprint import pprint

from sklearn.model_selection import train_test_split
from tqdm import tqdm
from transformers import BertTokenizerFast

import config
from common.utils import Preprocessor

config = config.train_config


def train_dev_test_split(semi=False):
    data = json.loads(unicodedata.normalize("NFKD", open('./data.json').read()))
    rs = []
    for line in data:
        if semi:
            rs.append({
                'id': line['id'],
                'text': line['text'],
                'relation_list': [{
                    "subject": attr['p'],
                    "object": attr['o'],
                    "predicate": '[INITIAL]',
                }
                    for attr in line['attribute'] if attr['o'] != ''
                ]
            })
        else:
            if line['s'] != '###' and line['s'] != '':
                try:
                    rs.append({
                        'id': line['id'],
                        'text': line['text'],
                        'relation_list': [{
                            "subject": line['s'],
                            "object": attr['o'],
                            "predicate": attr['p'],
                        }
                            for attr in line['attribute'] if attr['o'] != ''
                        ]
                    })
                except:
                    print('Error occured at id:', line['id'])
                    continue
    train_ratio = 0.75
    validation_ratio = 0.15
    test_ratio = 0.10

    train, test = train_test_split(rs, test_size=1 - train_ratio)
    val, test = train_test_split(test, test_size=test_ratio / (test_ratio + validation_ratio), shuffle=False)

    open('./ae_ori/train_data.json', 'w').write(json.dumps(train, ensure_ascii=False))
    open('./ae_ori/valid_data.json', 'w').write(json.dumps(val, ensure_ascii=False))
    open('./ae_ori/test_data.json', 'w').write(json.dumps(test, ensure_ascii=False))


def check_tok_span(data):
    def extr_ent(text, tok_span, tok2char_span):
        char_span_list = tok2char_span[tok_span[0]:tok_span[1]]
        char_span = (char_span_list[0][0], char_span_list[-1][1])
        decoded_ent = text[char_span[0]:char_span[1]]
        return decoded_ent

    span_error_memory = set()
    for sample in tqdm(data, desc="check tok spans"):
        text = sample["text"]
        tok2char_span = get_tok2char_span_map(text)

        for rel in sample["relation_list"]:
            subj_tok_span, obj_tok_span = rel["subj_tok_span"], rel["obj_tok_span"]
            if extr_ent(text, subj_tok_span, tok2char_span) != rel["subject"]:
                span_error_memory.add(
                    "extr: {}---gold: {}".format(extr_ent(text, subj_tok_span, tok2char_span), rel["subject"]))
            if extr_ent(text, obj_tok_span, tok2char_span) != rel["object"]:
                span_error_memory.add(
                    "extr: {}---gold: {}".format(extr_ent(text, obj_tok_span, tok2char_span), rel["object"]))

    return span_error_memory


tokenizer = BertTokenizerFast.from_pretrained(config["bert_path"], add_special_tokens=False, do_lower_case=False)
tokenize = tokenizer.tokenize
get_tok2char_span_map = lambda text: \
    tokenizer.encode_plus(text, return_offsets_mapping=True, add_special_tokens=False)["offset_mapping"]

preprocessor = Preprocessor(tokenize_func=tokenize,
                            get_tok2char_span_map_func=get_tok2char_span_map)


def process():
    data_in_dir = './ae_ori/'

    file_name2data = {}
    for path, folds, files in os.walk(data_in_dir):
        for file_name in files:
            file_path = os.path.join(path, file_name)
            file_name = re.match("(.*?)\.json", file_name).group(1)
            file_name2data[file_name] = json.load(open(file_path, "r", encoding="utf-8"))

    data_out_dir = './semiae/'

    rel_set = set()
    ent_set = set()
    error_statistics = {}
    for file_name, data in file_name2data.items():
        assert len(data) > 0
        if "relation_list" in data[0]:  # train or valid data
            data = preprocessor.clean_data_wo_span(data, separate=False)
            error_statistics[file_name] = {}
            # add char span
            data, miss_sample_list = preprocessor.add_char_span(data, ignore_subword_match=False)
            error_statistics[file_name]["miss_samples"] = len(miss_sample_list)

            #         # clean

            # collect relation types and entity types
            for sample in tqdm(data, desc="building relation type set and entity type set"):

                for rel in sample["relation_list"]:
                    # if rel['predicate'].startswith('@'):
                    rel_set.add(rel["predicate"])

            # add tok span
            data = preprocessor.add_tok_span(data)
            # check tok span
            span_error_memory = check_tok_span(data)
            if len(span_error_memory) > 0:
                print(span_error_memory)
            error_statistics[file_name]["tok_span_error"] = len(span_error_memory)

            file_name2data[file_name] = data
        pprint(error_statistics)

    rel_set = sorted(rel_set)
    # rel_set.append('[UNK]')  # Free properties
    rel2id = {rel: ind for ind, rel in enumerate(rel_set)}

    data_statistics = {
        "relation_type_num": len(rel2id),
    }

    for file_name, data in file_name2data.items():
        data_path = os.path.join(data_out_dir, "{}.json".format(file_name))
        json.dump(data, open(data_path, "w", encoding="utf-8"), ensure_ascii=False)
        logging.info("{} is output to {}".format(file_name, data_path))
        data_statistics[file_name] = len(data)

    rel2id_path = os.path.join(data_out_dir, "rel2id.json")
    json.dump(rel2id, open(rel2id_path, "w", encoding="utf-8"), ensure_ascii=False)
    logging.info("rel2id is output to {}".format(rel2id_path))

    data_statistics_path = os.path.join(data_out_dir, "data_statistics.txt")
    json.dump(data_statistics, open(data_statistics_path, "w", encoding="utf-8"), ensure_ascii=False, indent=4)
    logging.info("data_statistics is output to {}".format(data_statistics_path))

    pprint(data_statistics)


def main():
    train_dev_test_split(semi=True)
    process()


if __name__ == '__main__':
    main()

```

**Comment by githubgtl** [comment]
thank you very much！god bless you.



---Original---
From: ***@***.***&gt;
Date: Mon, Sep 26, 2022 14:35 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [lsvih/SOAE] there is no code, if you have, we had better learnyour inspired code (Issue #1)




 
Hi there, you can follow the setting from TPLinker and transfer our dataset to the corresponding data struct by referring to code below:
 import json import logging import os import re import unicodedata from pprint import pprint from sklearn.model_selection import train_test_split from tqdm import tqdm from transformers import BertTokenizerFast import config from common.utils import Preprocessor config = config.train_config def train_dev_test_split(semi=False):     data = json.loads(unicodedata.normalize("NFKD", open('./data.json').read()))     rs = []     for line in data:         if semi:             rs.append({                 'id': line['id'],                 'text': line['text'],                 'relation_list': [{                     "subject": attr['p'],                     "object": attr['o'],                     "predicate": '[INITIAL]',                 }                     for attr in line['attribute'] if attr['o'] != ''                 ]             })         else:             if line['s'] != '###' and line['s'] != '':                 try:                     rs.append({                         'id': line['id'],                         'text': line['text'],                         'relation_list': [{                             "subject": line['s'],                             "object": attr['o'],                             "predicate": attr['p'],                         }                             for attr in line['attribute'] if attr['o'] != ''                         ]                     })                 except:                     print('Error occured at id:', line['id'])                     continue     train_ratio = 0.75     validation_ratio = 0.15     test_ratio = 0.10     train, test = train_test_split(rs, test_size=1 - train_ratio)     val, test = train_test_split(test, test_size=test_ratio / (test_ratio + validation_ratio), shuffle=False)     open('./ae_ori/train_data.json', 'w').write(json.dumps(train, ensure_ascii=False))     open('./ae_ori/valid_data.json', 'w').write(json.dumps(val, ensure_ascii=False))     open('./ae_ori/test_data.json', 'w').write(json.dumps(test, ensure_ascii=False)) def check_tok_span(data):     def extr_ent(text, tok_span, tok2char_span):         char_span_list = tok2char_span[tok_span[0]:tok_span[1]]         char_span = (char_span_list[0][0], char_span_list[-1][1])         decoded_ent = text[char_span[0]:char_span[1]]         return decoded_ent     span_error_memory = set()     for sample in tqdm(data, desc="check tok spans"):         text = sample["text"]         tok2char_span = get_tok2char_span_map(text)         for rel in sample["relation_list"]:             subj_tok_span, obj_tok_span = rel["subj_tok_span"], rel["obj_tok_span"]             if extr_ent(text, subj_tok_span, tok2char_span) != rel["subject"]:                 span_error_memory.add(                     "extr: {}---gold: {}".format(extr_ent(text, subj_tok_span, tok2char_span), rel["subject"]))             if extr_ent(text, obj_tok_span, tok2char_span) != rel["object"]:                 span_error_memory.add(                     "extr: {}---gold: {}".format(extr_ent(text, obj_tok_span, tok2char_span), rel["object"]))     return span_error_memory tokenizer = BertTokenizerFast.from_pretrained(config["bert_path"], add_special_tokens=False, do_lower_case=False) tokenize = tokenizer.tokenize get_tok2char_span_map = lambda text: \     tokenizer.encode_plus(text, return_offsets_mapping=True, add_special_tokens=False)["offset_mapping"] preprocessor = Preprocessor(tokenize_func=tokenize,                             get_tok2char_span_map_func=get_tok2char_span_map) def process():     data_in_dir = './ae_ori/'     file_name2data = {}     for path, folds, files in os.walk(data_in_dir):         for file_name in files:             file_path = os.path.join(path, file_name)             file_name = re.match("(.*?)\.json", file_name).group(1)             file_name2data[file_name] = json.load(open(file_path, "r", encoding="utf-8"))     data_out_dir = './semiae/'     rel_set = set()     ent_set = set()     error_statistics = {}     for file_name, data in file_name2data.items():         assert len(data) &gt; 0         if "relation_list" in data[0]:  # train or valid data             data = preprocessor.clean_data_wo_span(data, separate=False)             error_statistics[file_name] = {}             # add char span             data, miss_sample_list = preprocessor.add_char_span(data, ignore_subword_match=False)             error_statistics[file_name]["miss_samples"] = len(miss_sample_list)             #         # clean             # collect relation types and entity types             for sample in tqdm(data, desc="building relation type set and entity type set"):                 for rel in sample["relation_list"]:                     # if rel['predicate'].startswith('@'):                     rel_set.add(rel["predicate"])             # add tok span             data = preprocessor.add_tok_span(data)             # check tok span             span_error_memory = check_tok_span(data)             if len(span_error_memory) &gt; 0:                 print(span_error_memory)             error_statistics[file_name]["tok_span_error"] = len(span_error_memory)             file_name2data[file_name] = data         pprint(error_statistics)     rel_set = sorted(rel_set)     # rel_set.append('[UNK]')  # Free properties     rel2id = {rel: ind for ind, rel in enumerate(rel_set)}     data_statistics = {         "relation_type_num": len(rel2id),     }     for file_name, data in file_name2data.items():         data_path = os.path.join(data_out_dir, "{}.json".format(file_name))         json.dump(data, open(data_path, "w", encoding="utf-8"), ensure_ascii=False)         logging.info("{} is output to {}".format(file_name, data_path))         data_statistics[file_name] = len(data)     rel2id_path = os.path.join(data_out_dir, "rel2id.json")     json.dump(rel2id, open(rel2id_path, "w", encoding="utf-8"), ensure_ascii=False)     logging.info("rel2id is output to {}".format(rel2id_path))     data_statistics_path = os.path.join(data_out_dir, "data_statistics.txt")     json.dump(data_statistics, open(data_statistics_path, "w", encoding="utf-8"), ensure_ascii=False, indent=4)     logging.info("data_statistics is output to {}".format(data_statistics_path))     pprint(data_statistics) def main():     train_dev_test_split(semi=True)     process() if __name__ == '__main__':     main()
